### ISSUE TITLE
Apply cross-language LTO to FxA client library

### ISSUE THREAD

**Issue by ncalexan** [mozilla/application-services#142]
Right now, the FxA client library built in https://github.com/mozilla-mobile/android-components/blob/79c917f05be971ba8387cc4d83c205b27c27c3d0/.taskcluster.yml packages itself as a .so and dynamically links against two other .so libraries: `libcrypto.so` and `libssl.so`.  If I understand the cross-language LTO story for Rust correctly, we should be able to build `libcrypto.a` and `libssl.a` and statically link _and LTO_ those libraries into `lib_fxa_rust.so`.  That means that symbols that we don't actually (transitively) use can be culled, which I expect to save a great deal of on-disk space.

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-70)

**Comment by eoger** [comment]
I assume on the iOS side of things https://github.com/rust-lang/rust/issues/35968 would help

**Comment by thomcc** [comment]
So, I made an attempt to megazord the FxA and Sync15Passwords libraries together a while ago. This is not really the same issue, but it's close enough and seems worth documenting the outcome (and i was requested that I do so in slack).

I ended up with a crate that basically contained just this https://gist.github.com/thomcc/596e93f851c9809965a83b0ee2b4cc98, and produced a library that was smaller than the sync15 passwords sofile alone (by a lot! Around 5MB IIRC). Analysis of the symbols exported showed that this prevented exporting of a bunch of pub mentat symbols that shouldn't be exported from either but who knows.

Basically, the naïve approach, just `pub use blah::*` possibly with a `#[no_mangle]` in front didn't quite do the trick. It turns out that in order to work on Android, the functions actually need to be referenced (and at the time I blamed https://github.com/rust-lang/rust/issues/36342, but in retrospect it worked on all platforms except for Android, so I think possibly something in the the GCC-based toolchain might really be responsible? or maybe it is 36342? I don't know). I tried a few tricks to get around this, there's one that worked but i find almost unconscionable (lines 15 to 60), and one that required changes to the fxa crate (to make sure stuff like SyncKeysC were `pub`), but was less terrible (all the rest of the code).

(This also took some other changes to the fxa ffi crate cargo.toml that I didn't preserve but they were pretty easy, I think just changing the crate type?)
